### PR TITLE
feat(metrics): metrics for PFD transactions, sampling, and broadcasts

### LINKED ISSUE #1040: [Feature Request]: metrics for PayForData and last sample requested
| Metric Name                              | Light node | Full Node | Bridge node |
| ---------------------------------------- | ---------- | ----------------- | ----------- |
| Total PayForData messages sent (count)   | x          | x                 | x           |
| Last PayForData message sent (timestamp) | x          | x                 | x           |
| Last sample requested (timestamp)        | x          | x                 |             |
| Last block served (timestamp)            |            |                   | x           |
| Last full sampling request (timestamp)   |            | x                 |             |

Feature request to emit the above metrics with a label that uniquely identifies a node (i.e. peer id). Note that a peer id is bounded on incentivized testnet but unbounded on mainnet so we likely want to remove the label prior to mainnet to avoid issues with high cardinality (see [Prometheus docs](https://prometheus.io/docs/practices/naming/#labels)).

### PR DESCRIPTION
**Pull request by distractedm1nd**
Closes #1040 

First time playing with metrics, so unsure if this is the right approach

| Metric Name                              | Light node | Full Node | Bridge node | Implemented |
| ---------------------------------------- | ---------- | ----------------- | ----------- | ----- |
| Total PayForData messages sent (count)   | x          | x                 | x           | ✅ |
| Last PayForData message sent (timestamp) | x          | x                 | x           | ✅ |
| Last sample requested (timestamp)        | x          | x                 |             | ✅ |
| Last block served (timestamp)            |            |                   | x           | ❌ |
| Last full sampling request (timestamp)   |            |                  | x           | ❌

- `Last full sampling req` is not implemented because it is the same as `Last sample requested`
- `Last block served` is not possible until we have implemented the networking of the new BlockSync
- All metrics reporters are now identifiable through their `peer.ID` 
- Tested on grafana cloud

I am marking as out of draft for reviews but this PR should not be merged until #997 is merged, because rebasing on this would take more time than this being rebased.

### CODE REVIEW

**Comment by distractedm1nd** [comment]
Started rebasing, but realized that it will be easier once #1125 is merged first - let's just make sure that the "Last sampled timestamp" makes it in there

**Comment by Wondertan** [comment]
Wanted to review this, but will wait for the #1125 to land first

File: nodebuilder/settings.go
Comment (renaynay) [inline nodebuilder/settings.go:52]: is there a reason to separate case node.Full and case node.Light? Unless it is expected to add diff metrics for full and light in the near future.

File: header/core/listener.go
Comment (renaynay) [inline header/core/listener.go:31]: i would call this `lastHeaderBroadcast` bc that implies you're measuring the broadcast time rather than the header height

File: nodebuilder/settings.go
Comment (renaynay) [inline nodebuilder/settings.go:104]:
```suggestion
	}
	
	global.SetMeterProvider(pusher)
	return nil
```

**Review by renaynay — COMMENTED** [review]
teeny nits

File: header/core/listener.go
Comment (distractedm1nd) [reply header/core/listener.go]: done

File: nodebuilder/settings.go
Comment (distractedm1nd) [reply nodebuilder/settings.go]: done

**Review by vgonkivs — APPROVED** [review]

File: service/state/metrics.go
Comment (Wondertan) [inline service/state/metrics.go:13]: What about `WithMetrics` so that all the modules on `nodebuilder` follow the same pattern? Even though it does PFD monitoring only, for now.

The main reason for `WithMetrics` because we are not planning to granular select which piece of the stack you want to monitor, so it's either `all or nothing` or `all for a module or nothing`.

File: header/core/listener.go
Comment (Wondertan) [inline header/core/listener.go:104]: Hmm, I don't really get why we would need this metric. It's also not part of the table.

File: service/state/metrics.go
Comment (distractedm1nd) [reply service/state/metrics.go:13]: Yes, agree

File: nodebuilder/settings.go
Comment (distractedm1nd) [reply nodebuilder/settings.go:96]: agreed

File: header/core/listener.go
Comment (distractedm1nd) [reply header/core/listener.go:104]: I can remove it if you want - source: you commented in #1040 that it wasnt in the table but would be useful to keep

Just respond here and i will either remove/resolve

Comment (Wondertan) [reply header/core/listener.go:104]: Quote from that issue:
 > The issue is about the last block served, not the header.
 
 Here we broadcast the header. Which IMO is an entirely useless metric.

Comment (distractedm1nd) [reply header/core/listener.go:104]: image

Comment (distractedm1nd) [reply header/core/listener.go:104]: will remove though, understood

File: nodebuilder/settings.go
Comment (distractedm1nd) [reply nodebuilder/settings.go]: done

File: service/state/metrics.go
Comment (distractedm1nd) [reply service/state/metrics.go]: done

File: header/core/listener.go
Comment (distractedm1nd) [reply header/core/listener.go]: done

**Review by renaynay — DISMISSED** [review]

File: nodebuilder/settings.go
Comment (Wondertan) [inline nodebuilder/settings.go:69]: Let's make it private and lfg

**Review by Wondertan — APPROVED** [review]

File: nodebuilder/settings.go
Comment (renaynay) [inline nodebuilder/settings.go:37]: Where is `WithMetrics` called that needs to pass in `false`? Let's remove this param?

Comment (renaynay) [reply nodebuilder/settings.go:37]: it just returns fx.Option so no need to short-circuit with the `enable` param

Comment (distractedm1nd) [reply nodebuilder/settings.go]: done

**Review by Wondertan — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Pull request merged**
